src/NSubstitute/Core/ThreadLocalContext.cs

```diff
@@ -54,6 +57,28 @@ public ConfiguredCall LastCallShouldReturn(IReturn value, MatchArgs matchArgs)
             return configuredCall;
         }
 
+        public void SetNextRoute(ICallRouter callRouter, Func<ISubstituteState, IRoute> nextRouteFactory)
+        {
+            if (callRouter == null) throw new ArgumentNullException(nameof(callRouter));
+            if (nextRouteFactory == null) throw new ArgumentNullException(nameof(nextRouteFactory));
+
+            _nextRouteFactory.Value = Tuple.Create(callRouter, nextRouteFactory);
+        }
+
+        public Func<ISubstituteState, IRoute> UseNextRoute(ICallRouter callRouter)
+        {
+            if (callRouter == null) throw new ArgumentNullException(nameof(callRouter));
+            
+            var value = _nextRouteFactory.Value;
+            if (value != null && value.Item1 == callRouter)
```

Comment:
I think I'd prefer to use `Object.ReferenceEquals(value.Item1, callRouter)` here to make the type of comparison more obvious.

Under what conditions can this comparison fail? (i.e. `UseNextRoute` called with a `callRouter` that is different to the one we have `SetNextRoute` for)

Reply:
Done. Thanks for the hint - looks much better with the `ReferenceEquals` API 👍 

> Under what conditions can this comparison fail?

That is for scenario described here:
https://github.com/nsubstitute/NSubstitute/blob/632a03c781df819221b73242e28ebff150b87931/tests/NSubstitute.Acceptance.Specs/ArgumentMatching.cs#L229-L240

Without this check `Received()` will be applied to the property getter. Check allows to ensure that other routers are not affected - exactly how it was before the change, where we stored route directly in the router.